### ISSUE TITLE
Layout feature proposal

### ISSUE THREAD

**Issue by abritinthebay** [jedireza/hapi-react-views#53]
(full disclosure, I'm writing this fork atm, would love to be able to issue a PR later to add it into the main stream)

My use case: I want the *entire* document tree (HTML, Head, etc) rendered by React on the server (controlled by a Store, etc) but this is problematic on the Client if you want to have the ability to embed third party scripts, analytics, etc. So you need to have the client render the core of your app somewhere in the Body.

If you render into a component other than HTML (when you're controlling the whole document on the server) then checksums will fail (assuming renderToStaticMarkup) because the `reactroot` is different when you render on the client - and if you don't use checksums then you're blitzing the node content which defeats some of the performance gains from having an Isomorphic app.

IDEALLY I'd have the ability to have a wrapper for whatever view I pass. This wrapper would always be use `rendertoString` and the wrapped component would use whatever was set in options (in my case renderToStaticMarkup). This is probably best handled in the options as a config.

Handling it in the options allows overriding it in the reply interface as well - which is very helpful.

an example:

`reply.view("myAppComponent", componentPropsObj, {wrapper: myHTMLComponent, wrapperProps: wrapperPropsObj})`

Thoughts on if this would be a desired optional feature? It certainly is *extremely* useful for fully controlled apps.

**Comment by jedireza** [comment]
Thanks for opening an issue. This is interesting. I'd like to think about it more, but this would be nice.

The usage should be clear; this is primariliy for universal/isomorphic applications. I'm not sure `wrapper` is telling that story.

I'd also like to investigate using the [built-in layout support](http://hapijs.com/tutorials/views#layouts) for this if at all possible. Then we could even pass a `layout` option during a `server.view(...)` call.

**Comment by abritinthebay** [comment]
While my use case is for isomorphic it could also be used for any generic wrappers that were wanted. I mean, obviously that's very applicable to isomorphic but it could be useful for just simple server-side. 

I'll play with the Layout system (I'd avoided it, but it might be useful for this). You're right - it might be a good fit.

**Comment by abritinthebay** [comment]
Ok, after further investigation the problem is the same as before unfortunately. While I can use layouts to render the Component I can't turn of renderToString for *just* that wrapper (as I want it for the rest). 

Therein lies the issue.

**Comment by abritinthebay** [comment]
One option would be to have the layout render option as a `compileOptions` key - IF I can work out when that is being rendered.

**Comment by abritinthebay** [comment]
After further investigation sadly Layouts does work for this. The layouts feature IS intended to potentially be bypassed by the engine though (and that's exactly what Jade does). So extending this way isn't unique or against the intent of how Vision works.

**Comment by abritinthebay** [comment]
For reference - this is basically the change (there's a little more but you can infer it from this). The Layout gets passed the context of the template as there is no way to pass on in a custom way (the `reply` interface isn't that flexible and the `Layout` functionality of Vision doesn't allow for setting the layout context dynamically).

```
		if (renderOpts.layout) {
			let Layout = require(renderOpts.layoutPath + renderOpts.layout);
			Layout = Layout.default || Layout;
			let layoutElement = React.createFactory(Layout);
			let layoutContext = Hoek.applyToDefaults({}, context);
			layoutContext[renderOpts.layoutKeyword] = PLACEHOLDER;

			let layoutOutput = ReactDOMServer[renderOpts.layoutRenderMethod](layoutElement(layoutContext));
			let elementOutput = ReactDOMServer[renderOpts.renderMethod](element(context));
			output = layoutOutput.replace(PLACEHOLDER, elementOutput);
		} else {
			output += ReactDOMServer[renderOpts.renderMethod](element(context));
		}```

if you can think of a nicer way to do it I'd LOVE to know. It does work though - and simply. Just requires a couple more render options.

**Comment by jedireza** [comment]
Let's setup a PR. I'd like to play with a working copy of this.

How are you handling routing on the client and the server?
How can we make this more configurable (like using a different render method for the wrapper/layout)?

**Comment by abritinthebay** [comment]
> How are you handling routing on the client and the server?

In my case it's all simply data driven (if property X is Y then load this component). Server side is the same (as it's the same store, which is then serialized). Obviously there's some routing via Hapi but really it's just to get the data for the app state (like, from url params/splats).

> How can we make this more configurable (like using a different render method for the wrapper/layout)?

Already in there :) see the line `let layoutOutput = ReactDOMServer[renderOpts.layoutRenderMethod](layoutElement(layoutContext));`

This means that I extend the defaults to add layout properties like so: 

```
const DEFAULTS = {
	doctype: "<!DOCTYPE html>",
	renderMethod: "renderToStaticMarkup",
	removeCache: process.env.NODE_ENV !== "production",
	layout: false,
	layoutPath: "",
	layoutKeyword: "children",
	layoutRenderMethod: "renderToStaticMarkup"
};
```

Children makes sense because then the default can just be `this.props.children` as is convention but it could be overridden to a specific prop if wanted. The `layoutRenderMethod` is the option you're specifically talking about.

These two code blocks are 99% of the changes btw ;)

**Comment by abritinthebay** [comment]
I'll slap a PR together either way, but thought I should detail the changes.

**Comment by jedireza** [comment]
So after further review of the pr in #54, I'm not convinced we've made anything better yet.

> My use case: I want the *entire* document tree (HTML, Head, etc) rendered by React on the server (controlled by a Store, etc) but this is problematic on the Client if you want to have the ability to embed third party scripts, analytics, etc. So you need to have the client render the core of your app somewhere in the Body.

We can do this today. I think handling layouts in the components is a more semantic/react way of approaching this. The [`examples/layout`](https://github.com/jedireza/hapi-react-views/tree/21148bb14d1c1e5e408713877211ec8203a1b2d0/examples/layout) demonstrates this. This is obviously best for static output that doesn't need to be remounted at the client (non universal/isomorphic).

> If you render into a component other than HTML (when you're controlling the whole document on the server) then checksums will fail (assuming renderToStaticMarkup) because the `reactroot` is different when you render on the client - and if you don't use checksums then you're blitzing the node content which defeats some of the performance gains from having an Isomorphic app.

Correct. The [`examples/remount`](https://github.com/jedireza/hapi-react-views/tree/21148bb14d1c1e5e408713877211ec8203a1b2d0/examples/remount) example shows how we can accomplish this and begin a universal/isomorphic journey. The main difference I see with the remount example and your proposal for layout options is that we save a nested call to `server.render(...)` to accomplish (roughly) the same thing.

There are many considerations when building universal/isomorphic applications; routing being a big one. I was hoping there was something related to these topics here but I'm not seeing it.

I'm going to close that PR, though that doesn't mean I'm not willing to discuss these ideas further.

👍 Thank you again for opening this issue and being involved.

**Comment by abritinthebay** [comment]
> This is obviously best for static output that doesn't need to be remounted at the client (non universal/isomorphic).

Right, which means for any dynamic use case (even setting a title or meta-tags dynamically) it's functionally _useless_, which means the statement:

> We can do this today.

Is actually not true. Not for the content most sites of any reasonable complexity need anyhow. SPA's may be ok.

> The examples/remount example shows how we can accomplish this 

Yes, as you said - it can be done, but by a very non-react idiom which loses most of the power of React at composition of components. The _one_ advantage of the remount example is independent contexts, which is nice, but that is completely outweighed by the complexity of the solution.

For something that renders `react-views` surely being "react-y" makes sense?

> There are many considerations when building universal/isomorphic applications; routing being a big one. I was hoping there was something related to these topics here but I'm not seeing it.

Routing is a function of how data interacts with your components - this is a completely separate issue. In fact I'd say it would be a mistake to have routing solutions in a view plugin - it's the wrong place for it. That said, if you expected that then I can understand your disappointment.

> I'm going to close that PR

Unfortunate, but I understand. I guess I'll have to fork - as this functionality and ease of programming is very important to my company's entire application stack. The remount example is just not worth implementing when there is a cleaner solution available.

I'm of course open to ways that the PR could be modified to fit in better with your idea, but I feel you're expecting something from it that it's not trying to do...

**Comment by jedireza** [comment]
> Right, which means for any dynamic use case (even setting a title or meta-tags dynamically) it's functionally useless, which means the statement:

But in the examples we're getting the title tag contents via the `title` attribute passed to the `Layout` component: https://github.com/jedireza/hapi-react-views/blob/21148bb14d1c1e5e408713877211ec8203a1b2d0/examples/layout/views/home.jsx#L9

That could be passed as a property to the component using `Layout`. But yes, not a great story for remounting.

> Not for the content most sites of any reasonable complexity need anyhow. SPA's may be ok.

This statement is vague. Would you be willing to clarify, maybe with specific examples?

> Yes, as you said - it can be done, but by a very non-react idiom which loses most of the power of React at composition of components. The _one_ advantage of the remount example is independent contexts, which is nice, but that is completely outweighed by the complexity of the solution.

> For something that renders `react-views` surely being "react-y" makes sense?

I don't follow. What is react-y about either approach? In your proposed changes we're just moving the functionality into the view engine instead of instrumenting it.

> That said, if you expected that then I can understand your disappointment.

I'm not disappointed. This has been interesting and fun. I got excited about a layout solution being proposed.

> this functionality and ease of programming is very important to my company's entire application stack.

I got this sense in the PR. The proposal comes with opinions that I've yet to be convinced apply well enough to the general use case.

I haven't written the whole idea off, I just closed that PR so we don't split the discussion away from this thread. It's possible that I'm not seeing all the use cases you have in mind.

**Comment by abritinthebay** [comment]
> I just closed that PR so we don't split the discussion away from this thread

Ah ok, gotcha. Wasn't sure of the goal there 😄 

> This statement is vague. Would you be willing to clarify, maybe with specific examples?

Sure. When you're build app state in a route it's going to *mostly* be specific to the request. Either the specific route OR a blend of that and the users request (user pages, specific article meta data, etc). Vision's layout capability isn't designed around this - your data is external, setup in the RenderOptions. It could be dynamic but it's never going to be specific to that route without using `Server.render`, which at that point *isn't using the layout engine*. So for any dynamic information in the layout that changes per-route you have to go with a much more verbose solution (as opposed to just `reply.view("myView", context);`) 

Now SPA's are likely to have simpler routing requirements so they may have less need to this (or can just go with some defaults and then override them later if needed). That's why I said they might be fine.

> What is react-y about either approach? In your proposed changes we're just moving the functionality into the view engine instead of instrumenting it.

Because it feels much more like standard Component Composition. A simple component nested inside another.

Compare just one simple line of code: `reply.view("component", context);` with the nested server render system in the remount example. Yes, the layout example in the PR requires a *one time* config setup when registering the plugin, but it requires dramatically less code from that point on.

In reality yes, it's actually doing two renders, just like the remount example (though slightly differently), but that is *abstracted away* from the developer. For applications with multiple routes this is a big deal.

>  The proposal comes with opinions that I've yet to be convinced applies well enough to the general use case.

You already have a remount example - this feature simply streamlines that functionality in a much more elegant syntax for isomorphic apps. If the remount functionality is a good general use case (i.e. - any isomorphic app) then so is this. It just makes it easier to use!

The opinions expressed are really just about adding syntax around the advanced layout-capabilities that are currently awkward to express in the current feature set.

I wish there was an elegant way to pass independent context with this system, but that said - the remount example *would still work* so it's not like it can't be done either. It's just a clean and simplified additional syntax.

**Comment by jedireza** [comment]
Resolved in #55
